// cdk-multi-profile-plugin/lib/ini-file-credential-provider-source.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IniFileCredentialProviderSource = void 0;
const safe_1 = require("colors/safe");
const lodash_isempty_1 = __importDefault(require("lodash.isempty"));
const aws_sdk_1 = require("aws-sdk");
const aws_cdk_1 = require("aws-cdk");
const utils_1 = require("./utils");
const profile_credentials_cache_1 = require("./profile-credentials-cache");
const profile_config_1 = require("./profile-config");
const sso_login_cache_1 = require("./sso-login-cache");
const profileCredentialsCache = new profile_credentials_cache_1.ProfileCredentialsCache();
class IniFileCredentialProviderSource {
    constructor(name, profiles, filename) {
        this.name = name;
        this.profiles = profiles;
        this.filename = filename;
        this.profileConfig = new profile_config_1.ProfileConfig(utils_1.getConfigFilename());
        this.ssoLoginCache = new sso_login_cache_1.SSOLoginCache(utils_1.getSSOCachePath());
    }
    canProvideCredentials(accountId) {
        return Promise.resolve(Object.prototype.hasOwnProperty.call(this.profiles, accountId));
    }
    async getProvider(accountId, mode) {
        const profile = this.profiles[accountId];
        console.log('\n');
        console.log(` 🚀  Using profile ${safe_1.green(profile)} for account ${safe_1.green(accountId)} in mode ${safe_1.green(aws_cdk_1.Mode[mode])}`);
        console.log('\n');
        let credentials = profileCredentialsCache.get(profile);
        if (!credentials) {
            if (this.profileConfig.isSSOProfile(profile)) {
                const ssoProfile = this.profileConfig.getProfile(profile);
                const ssoLogin = this.ssoLoginCache.getCachedLogin(ssoProfile);
                const sso = new aws_sdk_1.SSO({ region: ssoProfile.sso_region });
                const { roleCredentials } = await sso
                    .getRoleCredentials({
                    accessToken: ssoLogin.accessToken,
                    accountId: ssoProfile.sso_account_id,
                    roleName: ssoProfile.sso_role_name,
                })
                    .promise();
                if (!(roleCredentials === null || roleCredentials === void 0 ? void 0 : roleCredentials.accessKeyId) ||
                    !roleCredentials.secretAccessKey ||
                    !roleCredentials.sessionToken)
                    throw new Error('Invalid roleCredentials!');
                credentials = new aws_sdk_1.Credentials({
                    accessKeyId: roleCredentials.accessKeyId,
                    secretAccessKey: roleCredentials.secretAccessKey,
                    sessionToken: roleCredentials.sessionToken,
                });
            }
            else {
                credentials = new aws_sdk_1.SharedIniFileCredentials({
                    tokenCodeFn: utils_1.tokenCodeFn,
                    filename: this.filename,
                    profile,
                });
            }
            profileCredentialsCache.set(profile, credentials);
        }
        return Promise.resolve(credentials);
    }
    isAvailable() {
        if (this.filename && !lodash_isempty_1.default(this.profiles))
            return Promise.resolve(true);
        return Promise.resolve(false);
    }
}
exports.IniFileCredentialProviderSource = IniFileCredentialProviderSource;
