// cdk-multi-profile-plugin/lib/profile-mapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrecedenceProfileMapper = exports.LocalProjectDirMapper = exports.EnvironmentAwareGlobalProfileMapper = exports.PackageJsonProfileMapper = exports.JsonFileProfileMapper = void 0;
const fs = __importStar(require("fs-extra"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
/*
 * Resolves profiles from a json file. The key for resolving a mapping is "awsProfiles".
 *
 * example:
 *  `{
 *      "awsProfiles": {
 *        "123": "default123",
 *        "456": "default123"
 *      }
 *  }`
 *
 */
class JsonFileProfileMapper {
    constructor(props) {
        this._encoding = 'utf8';
        this._workingDirectory = props.workingDirectory;
        this._filename = props.filename;
    }
    resolve() {
        const filename = path.join(this._workingDirectory, this._filename);
        if (!fs.existsSync(filename)) {
            return {};
        }
        try {
            const pkg = JSON.parse(fs.readFileSync(filename, this._encoding));
            const { awsProfiles } = pkg;
            return awsProfiles;
        }
        catch (e) {
            console.log(`Failed to parse file ${this._filename}: `, e.message);
        }
        return {};
    }
}
exports.JsonFileProfileMapper = JsonFileProfileMapper;
// Will use local package.json for mapping an accountnumber to a local profile
class PackageJsonProfileMapper {
    resolve() {
        return new JsonFileProfileMapper({
            workingDirectory: process.cwd(),
            filename: 'package.json',
        }).resolve();
    }
}
exports.PackageJsonProfileMapper = PackageJsonProfileMapper;
// Will default to ~/.cdkmultiprofileplung.json and can be overriden by environment variable
// CDK_MULTI_PROFILE_PLUGIN_CONFIG=/path/to/file.json
class EnvironmentAwareGlobalProfileMapper {
    constructor() {
        this._defaultGlobalConfigurationFile = '.cdkmultiprofileplugin.json';
        this._workingDirectory = os.homedir();
        this._filename = this._defaultGlobalConfigurationFile;
        const configFileLocationOverride = process.env[EnvironmentAwareGlobalProfileMapper.environmentVariableName];
        if (configFileLocationOverride) {
            const configFile = path.parse(configFileLocationOverride);
            this._workingDirectory = configFile.dir;
            this._filename = configFile.base;
        }
    }
    resolve() {
        return new JsonFileProfileMapper({
            workingDirectory: this._workingDirectory,
            filename: this._filename,
        }).resolve();
    }
}
exports.EnvironmentAwareGlobalProfileMapper = EnvironmentAwareGlobalProfileMapper;
EnvironmentAwareGlobalProfileMapper.environmentVariableName = 'CDK_MULTI_PROFILE_PLUGIN_CONFIG';
// Can be used in local project directory. Can be added or ignored by your VCS
class LocalProjectDirMapper {
    resolve() {
        return new JsonFileProfileMapper({
            workingDirectory: process.cwd(),
            filename: 'cdkmultiprofileplugin.json',
        }).resolve();
    }
}
exports.LocalProjectDirMapper = LocalProjectDirMapper;
// Uses all mappers and applies precedence
class PrecedenceProfileMapper {
    resolve() {
        // Temporary in order to get it working again
        const packageJsonMappings = new PackageJsonProfileMapper().resolve();
        const projectLocalMappings = new LocalProjectDirMapper().resolve();
        const globalMappings = new EnvironmentAwareGlobalProfileMapper().resolve();
        return {
            ...packageJsonMappings,
            ...projectLocalMappings,
            ...globalMappings,
        };
    }
}
exports.PrecedenceProfileMapper = PrecedenceProfileMapper;
