// cdk-multi-profile-plugin/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSSOCachePath = exports.getConfigFilename = exports.getSharedCredentialsFilename = exports.readProfiles = exports.tokenCodeFn = void 0;
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const inquirer = __importStar(require("inquirer"));
const mfa_token_cache_1 = require("./mfa-token-cache");
const profile_mapper_1 = require("./profile-mapper");
const tokenCache = new mfa_token_cache_1.MfaTokenCache();
exports.tokenCodeFn = async (mfaSerial, callback) => {
    try {
        const { token } = await inquirer.prompt({
            name: 'token',
            type: 'input',
            default: '',
            message: `MFA token for ${mfaSerial}:`,
            validate: async (input) => {
                if (tokenCache.has(mfaSerial, input)) {
                    return `Token ${input} has already been used in this run`;
                }
                tokenCache.set(mfaSerial, input);
                return true;
            },
        });
        return callback(undefined, token);
    }
    catch (e) {
        console.error('error:', e);
        return callback(e, undefined);
    }
};
exports.readProfiles = () => {
    const profileMapper = new profile_mapper_1.PrecedenceProfileMapper();
    return profileMapper.resolve();
};
exports.getSharedCredentialsFilename = () => { var _a; return (_a = process.env.AWS_SHARED_CREDENTIALS_FILE) !== null && _a !== void 0 ? _a : path.join(os.homedir(), '.aws', 'credentials'); };
exports.getConfigFilename = () => { var _a; return (_a = process.env.AWS_CONFIG_FILE) !== null && _a !== void 0 ? _a : path.join(os.homedir(), '.aws', 'config'); };
exports.getSSOCachePath = () => path.join(os.homedir(), '.aws', 'sso', 'cache');
