// @mobileposse/auto-delete-bucket/dist/src/resource/auto-delete-bucket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_lambda_1 = require("@aws-cdk/aws-lambda");
const core_1 = require("@aws-cdk/core");
const aws_cloudformation_1 = require("@aws-cdk/aws-cloudformation");
const aws_s3_1 = require("@aws-cdk/aws-s3");
const path = require("path");
class AutoDeleteBucket extends aws_s3_1.Bucket {
    constructor(scope, id, props = {}) {
        super(scope, id, {
            ...props,
            removalPolicy: core_1.RemovalPolicy.DESTROY
        });
        const lambda = new aws_lambda_1.SingletonFunction(this, 'AutoBucketHandler', {
            uuid: '7677dc81-117d-41c0-b75b-db11cb84bb70',
            runtime: aws_lambda_1.Runtime.NODEJS_10_X,
            code: aws_lambda_1.Code.asset(path.join(__dirname, '../lambda')),
            handler: 'main.handler',
            lambdaPurpose: 'AutoBucket',
            timeout: core_1.Duration.minutes(15)
        });
        const provider = aws_cloudformation_1.CustomResourceProvider.lambda(lambda);
        // allow the bucket contents to be read and deleted by the lambda
        this.grantReadWrite(lambda);
        new aws_cloudformation_1.CustomResource(this, 'AutoBucket', {
            provider,
            resourceType: 'Custom::AutoDeleteBucket',
            properties: {
                bucketName: this.bucketName
            }
        });
    }
}
exports.AutoDeleteBucket = AutoDeleteBucket;
